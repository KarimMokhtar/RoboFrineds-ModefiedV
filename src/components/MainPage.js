import React, { useEffect } from "react";

import CardList from "../components/CardList";
import SearchBox from "../components/SearchBox";
import Scroll from "../components/Scroll";
import Header from "../components/Header";
import ErrorBoundry from "../components/ErrorBoundry";

class MainPage extends React.Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    this.props.loadRobots();
  }
  filterRobots = () => {
    return this.props.robots.filter((robot) =>
      robot.name.toLowerCase().includes(this.props.searchField.toLowerCase())
    );
  };
  render() {
    const { isPending, searchField, robots, handleSearch } = this.props;
    if (isPending) return <h1>Loading...</h1>;
    let searchedRobots = [];
    if (searchField.trim().length) {
        searchedRobots = this.filterRobots();
    } else {
      searchedRobots = [...robots];
    }
    return (
      <div className="tc">
        <Header />
        <SearchBox onSearchChange={handleSearch} />
        <Scroll>
          <ErrorBoundry>
            <CardList robots={searchedRobots} />
          </ErrorBoundry>
        </Scroll>
      </div>
    );
  }
}

export default MainPage;
