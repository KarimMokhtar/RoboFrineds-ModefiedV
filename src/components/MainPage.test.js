import React from 'react';
import MainPage from './MainPage';
import { shallow } from 'enzyme';

let wrapper;
beforeEach(() => {
  const mockProps = {
    loadRobots: jest.fn(),
    handleSearch: jest.fn(),
    robots: [],
    searchField: '',
    isPending: false,
  };
  wrapper = shallow(<MainPage {...mockProps} />);
});
describe('Main component tests', () => {
  expect.assertions(1);
  it('renders main page', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  expect.assertions(1);
  it('filters robots correctly', () => {
    expect(wrapper.instance().filterRobots()).toEqual([]);

    const mockFilterProps = {
      loadRobots: jest.fn(),
      handleSearch: jest.fn(),
      robots: [
        {
          id: 3,
          name: 'John',
          email: 'John@gmail.com',
        },
      ],
      searchField: 'j',
      isPending: false,
    };
    const wrapperFilter = shallow(<MainPage {...mockFilterProps} />);
    expect(wrapperFilter.instance().filterRobots()).toEqual([
      {
        id: 3,
        name: 'John',
        email: 'John@gmail.com',
      },
    ]);
  });

  expect.assertions(1);
  it('filters robots correctly2', () => {
    const mockFilterProps = {
      loadRobots: jest.fn(),
      handleSearch: jest.fn(),
      robots: [
        {
          id: 3,
          name: 'John',
          email: 'John@gmail.com',
        },
      ],
      searchField: 'a',
      isPending: false,
    };
    const filteredRobots = [];
    const wrapperFilter = shallow(<MainPage {...mockFilterProps} />);
    expect(wrapperFilter.instance().filterRobots()).toEqual(filteredRobots);
  });

  expect.assertions(1);
  it('filters robots pending', () => {
    const mockFilterProps = {
      loadRobots: jest.fn(),
      handleSearch: jest.fn(),
      robots: [
        {
          id: 3,
          name: 'John',
          email: 'John@gmail.com',
        },
      ],
      searchField: 'a',
      isPending: true,
    };
    const wrapperFilter = shallow(<MainPage {...mockFilterProps} />);
    expect(wrapperFilter.debug()).toMatchSnapshot();
  });
});
