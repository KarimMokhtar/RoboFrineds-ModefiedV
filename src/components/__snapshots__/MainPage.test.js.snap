// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component tests filters robots pending 1`] = `
"<h1>
  Loading...
</h1>"
`;

exports[`Main component tests renders main page 1`] = `
"<div className=\\"tc\\">
  <Memo() />
  <SearchBox onSearchChange={[Function: mockConstructor]} />
  <Scroll>
    <ErrorBoundary>
      <CardList robots={{...}} />
    </ErrorBoundary>
  </Scroll>
</div>"
`;
